nixos/nginx: add locations."name".uwsgiPass option and use it

SuperSandro2000 opened the pull request:
I am going to use that in another PR for searxng #346777 to reduce duplication for uwsgi.

This also allows to easily overwrite a uwsgi_pass target with an upstreams block.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SuperSandro2000 committed:
nixos/nginx: add locations."name".uwsgiPass and related options and use it

Ma27 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -442,6 +444,9 @@ let
         proxy_set_header Connection $connection_upgrade;
       ''}
+      ${optionalString (config.uwsgiPass != null)
+        "uwsgi_pass ${config.uwsgiPass};"
```
Do we also need the `proxyResolveWhileRunning` workaround for uwsgi?

SuperSandro2000 replied:
I think it functions the same but I don't think we need that option at all. According to the nginx docs, the record is cached for it's TTL lifetime. If you want to refresh it more often, I would recommend to reduce the records TTL to a minumum or use a caching proxy like unbound and set an upper limit on TTL.

I am going to create a PR to remove the option.

SuperSandro2000 replied:
#347164

Ma27 reviewed (commented):
So I haven't checked, but I assume you can't have both `proxyPass` and `uwsgiPass` in the same location?

If so, we'll probably want an assertion checking for this?

SuperSandro2000 commented:
> So I haven't checked, but I assume you can't have both `proxyPass` and `uwsgiPass` in the same location?
> 
> If so, we'll probably want an assertion checking for this?

Probably, I added one.

Ma27 commented:
> Probably, I added one.

I think you pushed that to the wrong PR.

SuperSandro2000 commented:
Pushed the assertion to this PR and added the if blocks for proxyResolveWhileRunning.

fpletz commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -1163,6 +1173,14 @@ in
+          Options services.nginx.service.virtualHosts.<name>.proxyPass and
+          services.nginx.virtualHosts.<name>.uwsgiPass are mutually exclusive.
+        '';
+      }
```
This assertion is causing the eval failure. This should recurse through the locations of a virtual host instead because those options are part of locations.

Ma27 commented:
@ofborg eval

Ma27 commented:
I'm a little surprised the `check-meta` check just shown an actual error. Also I don't see anything obvious in the change. cc @cole-h any idea what's up? Triggering eval didn't seem to help either here.

Ma27 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -442,6 +444,13 @@ let
+      ${optionalString (config.uwsgiPass != null && !cfg.proxyResolveWhileRunning)
+        "uwsgi_pass ${config.uwsgiPass};"
+      }
+      ${optionalString (config.uwsgiPass != null && cfg.proxyResolveWhileRunning) ''
```
I'm a little bit torn on this.

On one hand I think it's surprising that the `proxy*` options have an effect on uwsgi. OTOH it seems tedious to add counterparts for the uwsgi case.

I'm slightly leaning towards adding uwsgi counterparts for everything we need here. What do others think of this?

SuperSandro2000 replied:
I've added the ones that existed. uwsgi has no http_version or set_header commands.

fpletz replied:
I'm also leaning towards adding uwsgi counterparts because this behavior is not intuitive. If we keep the proposed behavior in this PR we IMHO should at least mention in the proxy-related options documentation that they also apply to uwsgi.

As far as I can see this only applies to:
  * `recommendedProxySettings`
  * `proxyResolveWhileRunning`
  * `proxyTimeout`

SuperSandro2000 replied:
Just done that :)

Ma27 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -442,6 +444,13 @@ let
+        "uwsgi_pass ${config.uwsgiPass};"
+      }
+      ${optionalString (config.uwsgiPass != null && cfg.proxyResolveWhileRunning) ''
+        set $nix_proxy_target "${config.proxyPass}";
```
This should be `uwsgiPass`.

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -94,14 +94,16 @@ let
+    ${command}        X-Forwarded-For $proxy_add_x_forwarded_for;
+    ${command}        X-Forwarded-Proto $scheme;
+    ${command}        X-Forwarded-Host $host;
+    ${command}        X-Forwarded-Server $host;
```
Good catch. I must admit I didn't e2e test this with my other branch since creation. 😓

SuperSandro2000 commented:
Just deployed this on top of #346777 and things are working as they should.

Ma27 commented:
Apologies, I have this tab open, will try to review soon.

Ma27 reviewed (changes requested):
Did you also test the combination with `recommendedUwsgiSettings`?
Because if I do so for mailman, I get

```
machine # [   53.922668] env[1344]: Traceback (most recent call last):
machine # [   53.923586] env[1344]:   File "/nix/store/p08j09lli1w0y8gk7j5mzvxd1m5v9nwj-python3-3.11.9-env/lib/python3.11/site-packages/django/core/handlers/wsgi.py", line 123, in __call__
machine # [   53.925868] env[1344]:     request = self.request_class(environ)
machine # [   53.926829] env[1344]:               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
machine # [   53.927734] env[1344]:   File "/nix/store/p08j09lli1w0y8gk7j5mzvxd1m5v9nwj-python3-3.11.9-env/lib/python3.11/site-packages/django/core/handlers/wsgi.py", line 71, in __init__
machine # [   53.929852] env[1344]:     self.method = environ["REQUEST_METHOD"].upper()
machine # [   53.930968] env[1344]:                   ~~~~~~~^^^^^^^^^^^^^^^^^^
machine # curl: (22) T[   53.933110] env[1344]: KeyError: 'REQUEST_METHOD'
```

By adding all uwsgi params by hand with

```
uwsgi_param  QUERY_STRING       $query_string;
uwsgi_param  REQUEST_METHOD     $request_method;
uwsgi_param  CONTENT_TYPE       $content_type;
uwsgi_param  CONTENT_LENGTH     $content_length;

uwsgi_param  REQUEST_URI        $request_uri;
uwsgi_param  PATH_INFO          $document_uri;
uwsgi_param  DOCUMENT_ROOT      $document_root;
uwsgi_param  SERVER_PROTOCOL    $server_protocol;
uwsgi_param  REQUEST_SCHEME     $scheme;
uwsgi_param  HTTPS              $https if_not_empty;

uwsgi_param  REMOTE_ADDR        $remote_addr;
uwsgi_param  REMOTE_PORT        $remote_port;
uwsgi_param  SERVER_PORT        $server_port;
uwsgi_param  SERVER_NAME        $server_name;
```

I fixed the problem (REQUEST_METHOD wasn't alone, I triggered a redirect loop then on a curl against `localhost`).

So apparently nginx somewhere sets those defaults and our declaration discards these settings?

I think this also means that the global `recommendedUwsgiSettings` isn't really functional since it gets overwritten by whatever level sets the uwsgi params I shared above.

SuperSandro2000 commented:
> Did you also test the combination with recommendedUwsgiSettings?

I think I forgot that. 🙈 When enabling it for searxng, I get a very similar error.

> So apparently nginx somewhere sets those defaults and our declaration discards these settings?

Yeah, we include ``include ${cfg.package}/conf/uwsgi_params;`` at a higher level and the uwsgi_param directives at the lower level completely overwrite that.

SuperSandro2000 commented:
So, I fixed a typo in ``HTTP_X_REAL-IP``, some missing ``;``, we close connections like we do for the normal proxy pass (inspired by the searx docs which pass through close for bot detection but I don't think we want to do that) and loaded ``conf/uwsgi_params`` everywhere we load the recommend values.

gdamjan commented:
> By adding all uwsgi params by hand with

yeah those are the nginx defaults:
https://github.com/nginx/nginx/blob/master/conf/uwsgi_params

> at a higher level and the uwsgi_param directives at the lower level completely overwrite that.

super stupid behaviour by nginx, but, you can do the `include` whenever `uwsgi_pass` is used?

gdamjan commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -94,14 +94,22 @@ let
     proxy_set_header        X-Forwarded-Server $host;
   '';
+  recommendedUwsgiConfig = pkgs.writeText "nginx-recommended-uwsgi_param-headers.conf" ''
+    uwsgi_param             HTTP_HOST $host;
+    uwsgi_param             HTTP_X_REAL_IP $remote_addr;
+    uwsgi_param             HTTP_X_FORWARDED_FOR $proxy_add_x_forwarded_for;
+    uwsgi_param             HTTP_X_FORWARDED_PROTO $scheme;
+    uwsgi_param             HTTP_X_FORWARDED_HOST $host;
+    uwsgi_param             HTTP_X_FORWARDED_SERVER $host;
```
I'm a bit confused why these are needed and if anything even would use them

gdamjan replied:
uwsgi_params are generally used by uwsgi and/or its plugins, not directly accessed by applications. Some of the *very specific* use-cases are here https://uwsgi-docs.readthedocs.io/en/latest/Nginx.html

I can't say the above params are recommendedUwsgiConfig

SuperSandro2000 replied:
Those are basically a mapping of the proxy_pass variables.

SuperSandro2000 commented:
> you can do the `include` whenever `uwsgi_pass` is used?

Yep

SuperSandro2000 commented:
@Ma27 how do we want to proceed here?

Ma27 commented:
I'll try to get to another review soonish :)

Ma27 committed:
nixos/nginx: remove custom HTTP_ header from recommendedUwsgiSettings

`uwsgi_pass` uses the uwsgi protocol instead of the HTTP protocol
(that's what `proxy_pass` is for). Also, the source IP is already
derived from the REMOTE_ADDR param that's specified by the uwsgi
defaults from nginx, similarily to how it's done for fcgi.

Hence, removing that since it seems not necessary (and perhaps even
wrong).

Ma27 reviewed (approved):
So, I think @gdamjan has a point here: uwsgi_pass isn't http, but uwsgi, so we don't have the issue that the http scheme, server etc. is wrong because it's just a proxied, internal request. I'm not even sure if it's correct in case of `X-Forwarded-For` since we don't have a changed client IP since nginx sets `REMOTE_ADDR` correctly.

To avoid another feedback loop, I decided to push a change to this branch that removes this portion. If you're OK with this @SuperSandro2000, I'd merge.

fpletz reviewed (approved):
Current state also LGTM. All variables that should be needed by default are shipped with nginx.
